src/Std/Time/Time/Unit/Nanosecond.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.UnitVal
+import Std.Time.Bounded
+import Std.Time.LessEq
+import Lean.Data.Rat
+import Std.Time.Time.Unit.Millisecond
+
+namespace Std
+namespace Time
+
+set_option linter.all true
+
+namespace Nanosecond
+
+/--
+`Ordinal` represents a bounded value for second, which ranges between 0 and 60.
+This accounts for potential leap second.
+-/
+def Ordinal := Bounded.LE 0 999999999
+  deriving Repr, BEq, LE, LT
+
+instance : OfNat Ordinal n where ofNat := Bounded.LE.ofFin (Fin.ofNat n)
+
+instance : Inhabited Ordinal where default := 0
+
+namespace Ordinal
+
+/--
+Convert to `Millisecond.Ordinal`
+-/
+def toMillisecond (nano : Ordinal) : Millisecond.Ordinal :=
+  nano.div 1000000 (by decide)
+
+/--
+Convert from `Millisecond.Ordinal`
+-/
+def ofMillisecond (nano : Millisecond.Ordinal) : Nanosecond.Ordinal :=
+  nano.mul_pos 1000000 (by decide)
+  |>.expandTop (by decide)
+
+end Ordinal
+
+/--
+`Offset` represents an offset in nanoseconds. It is defined as an `Int`.
+-/
+def Offset : Type := UnitVal (1 / 1000000000)
```

Comment:
I don't fully understand how all of the various ways to represent a timespan fit together. When are users expected to use one of the `Offset` types? When should they use `Duration`? How does `Date.Scalar` come into play here? If I have two dates, I might want to get the difference in days between them as a `Day.Offset`, or the specific amount of time between them as a `Duration`. Are both possible? I guess the latter only works if I have time zone information?

Generally, I think this is all really confusing. That is potentially fine given that time in general is very confusing. However, if we make the API very complex then we need to be really good with documentation. Ideally every docstring for one of these times should list all of the alternatives that might be a better fit for what the user is trying to achieve.

Reply:
I was trying to define `Date.Scalar` and `Time.Scalar` as wrappers around `Day.Offset` and `Nanoseconds.Offset` (not true right now). `Scalar` types represent absolute dates and times and `Offset` types represent shifts in dates and times, so they should be used in operations like `date.addDays`

Currently, `Time.Scalar` is defined as the same as `Duration`, which is designed for high precision using seconds as the primary unit and nanoseconds as a `Nanosecond.Span` but I think that I'm going to change `Time.Scalar` to use just `Nanosecond.Offset` or remove it.

I'm not sure if I should remove all the Scalar types or what should I do with them.

Reply:
Perhaps you could try to write an overview (from a user perspective) over the various types explaining what they represent and when to use which? I think a good example of this is the overview over the [java.time](https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/time/package-summary.html) package. Once we have that overview it might be easier to see which parts we might want to adjust.

Reply:
Added